### PR TITLE
feat: use css/auto as the default css mode

### PR DESCRIPTION
**Pull request by burhanuday**
### <samp>🤖 Generated by Copilot at 9912805</samp>

The pull request enables automatic detection of CSS modules based on the file extension, using the new `css/auto` module type. It also updates the default module rule and the tests to use this type for CSS files.

### <samp>🤖 Generated by Copilot at 9912805</samp>

* Replace the `CSS_MODULE_TYPE` constant with `CSS_MODULE_TYPE_AUTO` to enable automatic detection of CSS modules based on filename ([link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-c0078e4713a0e7a24dbc56d6dfda3cf139376325fc65ea70a8389d4d30b27f04L18-R18))
* Update the `cssRule` and `cssModulesRule` objects in `lib/config/defaults.js` to use the `CSS_MODULE_TYPE_AUTO` type instead of the previous types ([link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-c0078e4713a0e7a24dbc56d6dfda3cf139376325fc65ea70a8389d4d30b27f04L712-R712), [link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-c0078e4713a0e7a24dbc56d6dfda3cf139376325fc65ea70a8389d4d30b27f04L719-R719))
* Modify the `IS_MODULES` regular expression in `lib/css/CssParser.js` to allow hyphens in the file extension, to support CSS modules with extensions like `.module.scss` or `.module.less` ([link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-37ddc51f6c610706a74ec7aa006cbe834fb0d9ef9d69e58af824c5ac0f32a054L42-R42))
* Update the snapshot tests in `test/Defaults.unittest.js` to reflect the changes in the default module rules ([link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-2ca9058acdbb297c64353f1f82fdd9828c21911ce176861026c6a0bea93fb5c2L2245-R2245), [link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-2ca9058acdbb297c64353f1f82fdd9828c21911ce176861026c6a0bea93fb5c2L2252-R2252), [link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-2ca9058acdbb297c64353f1f82fdd9828c21911ce176861026c6a0bea93fb5c2L2262-R2262), [link](https://github.com/webpack/webpack/pull/17399/files?diff=unified&w=0#diff-2ca9058acdbb297c64353f1f82fdd9828c21911ce176861026c6a0bea93fb5c2L2270-R2270))

### CODE REVIEW

File: lib/config/defaults.js [hunk 1]
```diff
@@ -709,14 +709,14 @@ const applyModuleDefaults = (
 		}
 		if (css) {
 			const cssRule = {
-				type: CSS_MODULE_TYPE,
+				type: CSS_MODULE_TYPE_AUTO,
 				resolve: {
 					fullySpecified: true,
 					preferRelative: true
```
Comment (burhanuday) [inline lib/config/defaults.js:715]: @alexander-akait would you know we prefer relative paths for normal css files but not for css modules?

File: lib/css/CssParser.js [hunk 2]
```diff
@@ -39,7 +39,7 @@ const IMAGE_SET_FUNCTION = /^(-\w+-)?image-set$/i;
 const OPTIONALLY_VENDOR_PREFIXED_KEYFRAMES_AT_RULE = /^@(-\w+-)?keyframes$/;
 const OPTIONALLY_VENDOR_PREFIXED_ANIMATION_PROPERTY =
 	/^(-\w+-)?animation(-name)?$/i;
-const IS_MODULES = /\.module(s)?\.\w+$/i;
+const IS_MODULES = /\.module(s)?\.[\w-]+$/i;
```
Comment (burhanuday) [inline lib/css/CssParser.js:42]: changed regex to consider filenames like `styles.module.my-css`

File: test/__snapshots__/ConfigTestCases.basictest.js.snap [hunk 3]
```diff
@@ -1244,7 +1244,7 @@ body {
...
...\.css\\\\?1f16,\\\\.\\\\/anonymous-deep-nested\\\\.css\\\\?c0a8,\\\\.\\\\/layer-deep-deep-nested\\\\.css\\\\?4bce,\\\\.\\\\/layer-deep-nested\\\\.css\\\\?a03f,\\\\.\\\\/anonymous-nested\\\\.css\\\\?390d,\\\\.\\\\/media-deep-deep-nested\\\\.css\\\\?7047,\\\\.\\\\/style8\\\\.css\\\\?8af1,\\\\.\\\\/duplicate-nested\\\\.css,\\\\.\\\\/anonymous-deep-deep-nested\\\\.css\\\\?9cec,\\\\.\\\\/anonymous-deep-nested\\\\.css\\\\?dea4,\\\\.\\\\/layer-deep-deep-nested\\\\.css\\\\?4897,\\\\.\\\\/layer-deep-nested\\\\.css\\\\?4579,\\\\.\\\\/anonymous-nested\\\\.css\\\\?df05,\\\\.\\\\/all-deep-deep-nested\\\\.css\\\\?55ab,\\\\.\\\\/all-deep-nested\\\\.css\\\\?1513,\\\\.\\\\/all-nested\\\\.css\\\\?ccc9,\\\\.\\\\/style2\\\\.css\\\\?warning\\\\=6,\\\\.\\\\/style2\\\\.css\\\\?warning\\\\=7,\\\\.\\\\/style2\\\\.css\\\\?warning\\\\=8,\\\\.\\\\/style2\\\\.css\\\\?foo\\\\=unknown,\\\\.\\\\/style2\\\\.css\\\\?foo\\\\=unknown1,\\\\.\\\\/style2\\\\.css\\\\?foo\\\\=unknown2,\\\\.\\\\/style2\\\\.css\\\\?unknown3,\\\\.\\\\/style2\\\\.css\\\\?after-namespace,\\\\.\\\\/style2\\\\.css\\\\?multiple\\\\=1,\\\\.\\\\/style2\\\\.css\\\\?multiple\\\\=3,\\\\.\\\\/style2\\\\.css\\\\?strange\\\\=3,\\\\.\\\\/style\\\\.css;}",
```
Comment (burhanuday) [inline test/__snapshots__/ConfigTestCases.basictest.js.snap:1247]: change in deterministic id due to change in identity of module from `css/module` to `css/auto`

File: lib/css/CssParser.js (hunk 2 above)
Comment (alexander-akait) [reply lib/css/CssParser.js:42]: Let's add a test case for it, also let's change and use `.+` to handle all characters here

File: lib/config/defaults.js (hunk 1 above)
Comment (alexander-akait) [reply lib/config/defaults.js:715]: I think it is a bug

File: lib/config/defaults.js [hunk 4]
```diff
@@ -709,14 +709,14 @@ const applyModuleDefaults = (
...
 				}
 			};
 			const cssModulesRule = {
-				type: "css/module",
+				type: CSS_MODULE_TYPE_AUTO,
 				resolve: {
 					fullySpecified: true
 				}
```
Comment (alexander-akait) [inline lib/config/defaults.js:722]: We should rewrite our logic:
```
const cssRule = {
  type: CSS_MODULE_TYPE,
  resolve: {
    fullySpecified: true,
    preferRelative: true
  }
};
const cssRuleAuto = {
  type: CSS_MODULE_TYPE_AUTO,
  resolve: {
    fullySpecified: true,
    preferRelative: true
  }
};
const cssModulesRule = {
  type: CSS_MODULE_TYPE_AUTO,
  resolve: {
    fullySpecified: true
  }
};
// Handle all CSS, and we will check is it modules inside parser
rules.push({
  test: /\.css$/i,
  ...cssRule,
});
rules.push({
  mimetype: "text/css+module",
  // Should be always only modules
  ...cssModulesRule
});
rules.push({
  mimetype: "text/css",
  // Should be always pure CSS
  ...cssRule
});
```

And let's add a test case for the each rule (for some we already have)

File: lib/css/CssParser.js (hunk 2 above)
Comment (burhanuday) [reply lib/css/CssParser.js:42]: Changing to `.+` will also match `some.module.abc.abc` which is invalid. In the future, we need to check the extensions array and match based on that

Comment (alexander-akait) [reply lib/css/CssParser.js:42]: Just use `[^.]+`

Comment (alexander-akait) [reply lib/css/CssParser.js:42]: We should avoid parsing extensions - it is slow

**Review by angeljavi — CHANGES_REQUESTED** [review]
No entiendo inglés y no sé leerlos

File: test/configCases/css/css-modules/use-style.js [hunk 5]
```diff
@@ -1,5 +1,7 @@
 import * as style from "./style.module.css";
 import { local1, local2, local3, local4, ident } from "./style.module.css";
+import { myCssClass } from "./style.module.my-css";
+import { notACssModule } from "./style.module.css.invalid"
```
[the remark this answers is not in the document]
Comment (alexander-akait) [reply test/configCases/css/css-modules/use-style.js:4]: I will look today

**Review by alexander-akait — APPROVED** [review]
Fixed, we have:
- bug, because we cache parser we need to reset defaultMode
- also added test cases and fixed preferRelative

**Review by TheLarkInn — APPROVED** [review]

**Pull request merged**